### ISSUE TITLE
HRTF support (OpenAL Soft Backend?)

### ISSUE THREAD

**Issue by Hiradur** [mumble-voip/mumble#1933]
I read on the Mumble wiki about  [HRTF](https://wiki.mumble.info/wiki/Projects#HRTF) and I wonder if the implementation of an OpenAL Soft backend has ever been considered. OpenAL Soft can process HRTF calculations internally.
It's fairly easy to achieve HRTF processing with OAL Soft, all you need is an HRTF data set (OAL Soft ships with a default one), set the listener's Position, At and Up vector correctly (Mumble already has this data through Mumble Link), place the player voices correctly (Mumble receives that data from other clients) and turn HRTF processing on.

Benefits of OpenAL Soft:
- FOSS (LGPLv2)
- Cross platform (both in terms of OS and CPU architecture)
- internal calculation of HRTF
  - can be requested by app or enforced by user for all OpenAL apps
  - calculation is optimized: intrinsics for NEON, SSE and AVX instruction sets
  - over 50 compatible HRTF data sets to find one which matches your head shape best
- internal calculation of Doppler Effect (if desired)
- OpenAL EFX to simulate reverbation of environment (if desired)
  - example video: https://www.youtube.com/watch?v=30fTc5t5QNU&fmt=18 (EFX is the successor of EAX)

Links:
Official site of OpenAL Soft: http://kcat.strangesoft.net/openal.html
OpenAL Soft repo on GitHub: https://github.com/kcat/openal-soft
Demonstration video of OpenAL Soft HRTF: https://www.youtube.com/watch?v=UtFatWeEvGY

**Comment by bontibon** [comment]
I used OpenAL when I made [my terminal Mumble client](https://github.com/layeh/barnard), and it worked great.

From the OpenAL soft project page:

> ALSA, OSS, DirectSound, PulseAudio, MMDevAPI, CoreAudio, Solaris, QSA, SoundIO, OpenSL, WinMM, PortAudio, "Null" Output, and a .wav writer are currently implemented.

@Hiradur Were you thinking that OpenAL replace all of Mumble's current audio systems, or just have it as an additional one? It seems like it *could* replace everything, but someone who knows the audio system better could probably tell me why this is a bad idea.

**Comment by Hiradur** [comment]
@bontibon Very nice, I'll check it out. 
For now I only propose to use OpenAL Soft as an additional backend not replacing the existing ones. I'm unfamiliar with how OpenAL handles input capturing and low level technical details so I can't say if it's suited to replace the existing backends or not.

**Comment by mkrautz** [comment]
In general, we are never opposed to new audio backends, as long as there is someone to maintain them.

We already have problems in this regard with PulseAudio/ALSA.

**Comment by Hiradur** [comment]
@bontibon Does barnard have support for the positional audio feature?

**Comment by DGMurdockIII** [comment]
Any progress on this

**Comment by perfectcolors** [comment]
I agree, an optional OpenAL Soft backend would be really nice here for Mumble Link supported games in order to take advantage of the built-in HRTF processing.

**Comment by foresto** [comment]
Just adding some keywords here to make it easier to find now that the HRTF discussion in #2324 (simple binaural audio) has ended:

It would be impressive if mumble could do advanced positional audio with a [head-related transfer function](https://en.wikipedia.org/wiki/Head-related_transfer_function). The experience in Overwatch, where I could hear when someone was (for example) above and behind my left shoulder so reliably that I could swivel around to aim at them in a single motion, was excellent.

**Comment by mirh** [comment]
I'm pretty sure openal-soft could actually allow do ditch away all the [OS-specific](https://github.com/kcat/openal-soft/tree/master/alc/backends) cruft, if one really wanted. 
And to the extent of my [understanding](https://www.pcgamingwiki.com/wiki/Glossary:Surround_sound#Headphones), there really isn't any other cross-platform HRTF implementation (let alone open source)

On the other hand, from the point of view of an application, you actually simply just code against "OpenAL".. not openal-soft specifically.

**Comment by mirh** [comment]
https://ieeexplore.ieee.org/document/5661988
Did nobody know about this?
@jfeldmaier

**Comment by davidebeatrici** [comment]
Yes, see https://github.com/mumble-voip/mumble/issues/2324#issuecomment-226331887.

I now uploaded the PDF file on our server: https://dl.mumble.info/docs/rothbucher2010.pdf

**Comment by Krzmbrzl** [comment]
I have created a (hopefully) more streamlined issue that tracks positional audio improvements via advanced physical sound modelling at #6597.
Future discussion about this feature should be directed to that issue.

The main reason for creating a new issue is that I have tried to summarize this feature as well as some other feature requests in the same spirit in the hope to make these things better findable via the issue search.
